sql/core/src/main/scala/org/apache/spark/sql/functions.scala

```diff
@@ -652,6 +652,50 @@ object functions {
    */
   def min(columnName: String): Column = min(Column(columnName))
 
+  /**
+   * Aggregate function: Returns and array of the approximate percentile values
+   * of numeric column col at the given percentages.
+   *
+   * Each value of the percentage array must be between 0.0 and 1.0.
+   *
+   * The accuracy parameter is a positive numeric literal
+   * which controls approximation accuracy at the cost of memory.
+   * Higher value of accuracy yields better accuracy, 1.0/accuracy
+   * is the relative error of the approximation.
+   *
+   * @group agg_funcs
+   * @since 3.1.0
+   */
+  def percentile_approx(e: Column, percentage: Array[Double], accuracy: Long): Column = {
```

Comment:
@zero323, about https://github.com/apache/spark/pull/27278#discussion_r368780883, let's just match to `Column` for now, merge it, and separately discuss. This signature thing is a kind of guideline we made to keep this file consistent. Changing this will require separate discussion including people who I also talked with.

For Python list issue,

```python
>>> from pyspark.sql.functions import lit
>>> import array
>>> spark.range(1).select(lit(["a", "b"]))
```

let's also fix it in a separate PR:

Reply:
Huh... Usability aside, could you clarify one thing for me. Percentiles have to be provided, as a literal. Even assuming that prevailing attitude is "exception is good enough" when non literal is passed

> AnalysisException: cannot resolve ... due to data type mismatch: The accuracy or percentage provided must be a constant literal;

However, arguments that require SQL literals:

- `ignoreNulls` in `first`
- `ignoreNulls` in `last`
- ` rsd`in `approx_count_distinct` 

not to count different args in strict window functions are exclusively passed as external literals not `Columns`\*. So what is the rationale behind diverging here? What makes this case different? 

It doesn't seem to be particularly verbose, two signatures are on the low side for overloaded functions with multiple arguments  and I don't see any reason to anticipate significant maintenance effort (docs or otherwise).

I've seen some echos of discussions about growing size of `o.a.s.sql.functions`, but these were hardly convincing, at least in the diluted form they reached me.

---------------------------

* In fact there are quite a few functions, where only external literal are passed, despite planner accepting any expression.

Reply:
(and yes, I know I repeat myself here, and I am really sorry for that. I really appreciate your reviews @HyukjinKwon, it's just some of these unwritten rules seem rather inconsistent and arbitrary for a passer-by like me, and navigating around these is not easy).

Reply:
Seems like the percentiles can at least be foldable expressions, e.g.:

```scala
scala> sql("select percentile_approx(1, 0.5 + 0.5)")
res5: org.apache.spark.sql.DataFrame = [percentile_approx(1, CAST((CAST(0.5 AS DECIMAL(2,1)) + CAST(0.5 AS DECIMAL(2,1))) AS DOUBLE), 10000): int]
```

I know it's pretty unlikely but still possible given that we have a set of foldable expressions. The error message isn't completely correct because it can accept other foldable expressions.

Other variants were not changed only due to the legacy reason.

I think we would have not had such arguments all signatures were with `Column` at the very first place in `functions.scala`.

Reply:
> I know it's pretty unlikely but still possible given that we have a set of foldable expressions. 

Yeah, it don't see that happening, though SQL folks can do all kinds of crazy stuff :)

> Other variants were not changed only due to the legacy reason.

Fair enough. Not what I wanted to hear, but what one can do? Thank you taking your time.

> I think we would have not had such arguments all signatures were with `Column` at the very first place in `functions.scala`.

That wouldn't be bad. If everything took only columns, it would be actually possible to add implicit conversions from common types to literals.

And to be honest Scala users are more likely to use IDE which can make sense out of things, not so much Python user and for sure, not R users.


> Also, it's completely no problem for any one in the community to say out what they think is right .. :-). We just need to chime in which I think I am not pretty good at .

I always value your input. My problem is that I don't have contact with Spark, unless something goes wrong (could be actual bug, could be users frustrated by the API, to the point, where they just give up).  As you can imagine that's creates highly biased view of UX.


I'll take another look at this, but at the moment I don't have environment that can handle Spark, so it will take some time.